File: src/main/java/GenerateParentheses/BBIGBOSS/GenerateParentheses.cpp
//
// Created by BBIGBOSS on 2019-06-15.
//

class Solution
{
    public:

    vector<string> result;

    vector<string> generateParenthesis(int n)
    {
        core(0, 0, n, "");
        return result;
    }

    void core(int left, int right, int n, string s)
    {
        if (right == n)
        {
            result.emplace_back(s);
            return;
        }

        if (left < n)
            core(left + 1, right, n, s + "(");

        if (right < left)
            core(left, right + 1, n, s + ")");
    }
};

File: src/main/java/GenerateParentheses/GGOBOND/Generate_Parentheses.py
class Solution:
    def generateParenthesis(self, n: int

    ) -> List[str]:
    res = []
    self.generate(n, n, '', res)
    return res


def generate(self, left, right, str_p, res):
    if left == 0 and right == 0:
        res.append(str_p)
        return

    if left > 0:
        self.generate(left - 1, right, str_p + '(', res)

    if right > left:  ### 此处是left > right   #if right > 0:
        self.generate(left, right - 1, str_p + ')', res)